### PR TITLE
[SPARK-48650][PYTHON] Display correct call site from IPython Notebook

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to display correct call site information from IPython Notebook.

### Why are the changes needed?

We added `DataFrameQueryContext` for PySpark error message from https://github.com/apache/spark/pull/45377, but it does not working very well from IPython Notebook.

### Does this PR introduce _any_ user-facing change?

No API changes, but the user-facing error message from IPython Notebook will be improved:

**Before**
Screenshot 2024-06-18 at 5 15 56 PM

**After**
Screenshot 2024-06-19 at 8 45 05 AM

**NOTE:** This also works when command is executed across multiple cells:

Screenshot 2024-06-19 at 8 42 29 AM

### How was this patch tested?

Manually tested with IPython Notebook.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by allisonwang-db — COMMENTED** [review]
I feel the file name doesn't make much sense, but the line number really matters. Just a suggestion: is there a way to highlight the line number?

**Comment by itholic** [comment]
@allisonwang-db Actually that is a good point. Let me try highlighting and providing more useful debugging information from error message.

**Comment by itholic** [comment]
Now we use the cell number instead of file name for better debuggability.

Updated before/after in PR description. cc @allisonwang-db FYI

File: python/pyspark/errors/utils.py [hunk 1]
```diff
@@ -165,8 +165,37 @@ def _capture_call_site(spark_session: "SparkSession", depth: int) -> str:
...
+    if ipython:
+        import pyspark
+
+        # Filtering out PySpark code and keeping user code only
+        pyspark_root = os.path.dirname(pyspark.__file__)
+        stack = [
+            frame_info for frame_info in inspect.stack() if pyspark_root not in frame_info.filename
+        ]
+
+    depth = int(
```
[the remark this answers is not in the document]
Comment (itholic) [reply python/pyspark/errors/utils.py:187]: I think maybe we already respect `depth` below by `selected_frames = stack[:depth]` ??

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/pyspark/errors/utils.py:187]: now this functions take a parameter called `depth` :-).

[the remark this answers is not in the document]
Comment (itholic) [reply python/pyspark/errors/utils.py:187]: Oh, didn't realize it. Let me address

[the remark this answers is not in the document]
Comment (itholic) [reply python/pyspark/errors/utils.py]: adjusted!

**Review by HyukjinKwon — APPROVED** [review]

**Review by allisonwang-db — APPROVED** [review]
This looks much better!

**Review by ueshin — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**